src/oph/ehoks/hoks/handler.clj

```diff
@@ -284,28 +287,71 @@
                (pdb-ot/select-opiskeluvalmiuksia-tukevat-opinnot-by-id id)}))
         (response/not-found {:error "OTO not found with given OTO ID"})))))
 
-(defn- post-hoks!
+(def ^:private tuva-hoks-msg-template
+  "HOKS `%s` is a TUVA-HOKS or rinnakkainen ammatillinen HOKS.")
+
+(defn- initiate-all-palautteet!
+  [{:keys [hoks] :as ctx}]
+  (if (hoks/tuva-related? hoks)
+    (db-hoks/set-amisherate-kasittelytilat-to-true!
+      (:id hoks) (format tuva-hoks-msg-template (:id hoks)))
+    (try
+      (op/initiate-if-needed! ctx :aloituskysely)
+      (op/initiate-if-needed! ctx :paattokysely)
+      (tep/initiate-all-uninitiated! ctx)
+      (catch clojure.lang.ExceptionInfo e
+        (if (= :organisaatio/organisation-not-found (:type (ex-data e)))
+          (throw (ex-info (str "HOKS contains an unknown organisation"
+                               (:organisation-oid (ex-data e)))
+                          (assoc (ex-data e) :type ::disallowed-update)))
+          (log/error e "exception in heräte initiation with" (ex-data e))))
+      (catch Exception e
+        (log/error e "exception in heräte initiation")))))
+
+(defn save-hoks-and-initiate-all-palautteet!
+  "Saves a HOKS to DB and initializes all palautteet (opiskelija & tyoelama)
+  that need to be initialized."
+  [{:keys [hoks] :as ctx}]
+  (jdbc/with-db-transaction
+    [tx db/spec]
+    (let [hoks (assoc hoks :id (:id (hoks/save! hoks)))]
+      (initiate-all-palautteet! (assoc ctx :hoks hoks :tx tx))
+      hoks)))
+
+(defn change-hoks-and-initiate-all-palautteet!
+  "Updates a HOKS in DB with `db-handler` (either update or replace) and
+  initializes all palautteet (opiskelija & tyoelama) that need to be
+  initialized."
+  [{:keys [hoks] :as ctx} db-handler]
+  (jdbc/with-db-transaction
+    [tx db/spec]
+    (let [updated-hoks (db-handler hoks)]
+      (initiate-all-palautteet! (assoc ctx :hoks updated-hoks :tx tx))
+      updated-hoks)))
```

Comment:
Äkkiseltään näyttäis että nää vois melko helposti yhdistää yhdeksi funktioksi, jota ottaa tuon db-handler-tallennus/päivitys-funktion (POST-requestin tapauksessa hoks/save!), mutta tuon tekeminen ei ollutkaan ihan niin suoraviivaista 😅